feat(manager/crossplane): add Crossplane xpkgs support (#25896)
renovatebot/renovate: lib/modules/manager/crossplane/readme.md

Reviewer: document possible `depType`s

> The `crossplane` manager has no `fileMatch` default patterns, so it won't match any files until you configure it with a pattern.

Reviewer 2: ```suggestion
To use the `crossplane` manager you must set your own `fileMatch` pattern.
This is because the `crossplane` manager has no default patterns for the `fileMatch`.
The `crossplane` manager only matches files when you set a `fileMatch` pattern.
```

We should use the word _must_ to indicate requirements. 😉

> This is because there is no commonly accepted file/directory naming convention for crossplane YAML files and we don't want to check every single `*.yaml` file in repositories just in case any of them have Crossplane Packages definitions: Configurations, Providers, Functions.

Reviewer 2: ```suggestion

There is no commonly accepted file/directory naming convention for crossplane YAML files.
We don't want Renovate to check each `*.yaml` file in repositories just in case a files has one of these Crossplane Package definitions:

- Configurations
- Providers
- Functions
```

If most `.yaml` files in your repository are Crossplane ones, then you could add this to your config:

> ```json

Reviewer 2: ```
json title="fileMatch if most .yaml files are for Crossplane"
```

First read my _other_ comment about this config example. Then fix up the title if needed.

{
  "crossplane": {
    "fileMatch": ["\\.yaml$"]
  }
}

> ```

Reviewer 2: I don't know if "If most `.yaml files are Crossplane files" is good enough. Let's say you use GitHub Actions in your repository, those use `.yaml` as config language. When you set the `fileMatch` for your Crossplane files, I supect Renovate will mess up the GitHub Actions `.yaml` files when it updates?

I suspect that only when _all_ `.yaml` files are Crossplane files it's a decent idea to set the `fileMatch` like in the example?

If instead you have them all inside a `packages/` directory, you would add this:

> ```json

Reviewer 2: ```
json title="fileMatch for Crossplane files in a packages/ directory"
```

{
  "crossplane": {
    "fileMatch": ["packages/.+\\.yaml$"]
  }
}
```

Or if it's only a single file then something like this:

> ```json

Reviewer 2: Can you add a codeblock title, and drop the introductory sentences?

```
json title="fileMatch for a single Crossplane file"
```

[Material for MkDocs manual, adding a title to codeblocks](https://squidfunk.github.io/mkdocs-material/reference/code-blocks/#adding-a-title)

{
  "crossplane": {
    "fileMatch": ["^config/provider\\.yaml$"]
  }
}
```

> If you need to change the versioning format, read the [versioning](https://docs.renovatebot.com/modules/versioning/) documentation to learn more.

Reviewer 2: This should be a relative link.

> The `crossplane` manager has three `depType`s to allow a fine-grained control of which dependencies are upgraded:
> 
> - `configuration`
> - `function`
> - `provider`

Reviewer 2: ```suggestion
The `crossplane` manager has three `depType`s:

- `configuration`
- `function`
- `provider`

You can use these `depType`'s to control which dependencies Renovate will upgrade.
```
